libcontainer/cgroups/devices/devices_emulator.go

```diff
@@ -79,19 +78,24 @@ func (e *Emulator) IsAllowAll() bool {
 	return e.IsBlacklist() && len(e.rules) == 0
 }
 
-var devicesListRegexp = regexp.MustCompile(`^([abc])\s+(\d+|\*):(\d+|\*)\s+([rwm]+)$`)
-
 func parseLine(line string) (*deviceRule, error) {
-	matches := devicesListRegexp.FindStringSubmatch(line)
-	if matches == nil {
-		return nil, errors.New("line doesn't match devices.list format")
+	// The three fields are: node major:minor perms.
+	fields := strings.Fields(line)
```

Comment:
Ideally, I'd use `strings.Split(line, " ")` here -- it works just fine as contents from the kernel do not have any extra spaces, and makes is 1.5x faster. OTOH `strings.Fields` is safer, as it will eat extra whitespace. I'm on the verge here -- WDYT @thaJeztah ?

Reply:
I'm trying to Benchmark the difference, but for some reason, I don't seem to be doing things right (it shows `strings.Fields()` to be actually _faster_ ? I was considering `strings.FieldsFunc()`, as it would allow splitting on spaces _only_ (not tabs, newlines, etc), and still keeping the advantage of `strings.Fields()` (trimming whitespace)

I tried `strings.Fields()`, `strings.FieldsFunc()`, and `strings.Split()`

<details>
    <summary>test_test.go</summary>
    
    ```go
    package main
    
    import (
    	"bufio"
    	"strings"
    	"testing"
    )
    
    
    func BenchmarkParseFields(b *testing.B) {
    	list := `c *:* m
    b *:* m
    c 1:3 rwm
    c 1:5 rwm
    c 1:7 rwm
    c 1:8 rwm
    c 1:9 rwm
    c 5:0 rwm
    c 5:2 rwm
    c 136:* rwm
    c 10:200 rwm`
    
    	var out []string
    	for i := 0; i < b.N; i++ {
    		s := bufio.NewScanner(strings.NewReader(list))
    		for s.Scan() {
    			line := s.Text()
    			out = strings.Fields(line)
    		}
    		if err := s.Err(); err != nil {
    			b.Fatal(err)
    		}
    	}
    	b.Logf("rule: %v", out)
    }
    
    func isSpace(r rune) bool {
    	return r == ' '
    }
    
    func BenchmarkParseFieldFunc(b *testing.B) {
    	list := `c *:* m
    b *:* m
    c 1:3 rwm
    c 1:5 rwm
    c 1:7 rwm
    c 1:8 rwm
    c 1:9 rwm
    c 5:0 rwm
    c 5:2 rwm
    c 136:* rwm
    c 10:200 rwm`
    
    	var out []string
    	for i := 0; i < b.N; i++ {
    		s := bufio.NewScanner(strings.NewReader(list))
    		for s.Scan() {
    			line := s.Text()
    			out = strings.FieldsFunc(line, isSpace)
    		}
    		if err := s.Err(); err != nil {
    			b.Fatal(err)
    		}
    	}
    	b.Logf("rule: %v", out)
    }
    
    func BenchmarkParseSplit(b *testing.B) {
    	list := `c *:* m
    b *:* m
    c 1:3 rwm
    c 1:5 rwm
    c 1:7 rwm
    c 1:8 rwm
    c 1:9 rwm
    c 5:0 rwm
    c 5:2 rwm
    c 136:* rwm
    c 10:200 rwm`
    
    	var out []string
    	for i := 0; i < b.N; i++ {
    		s := bufio.NewScanner(strings.NewReader(list))
    		for s.Scan() {
    			line := s.Text()
    			out = strings.Split(line, " ")
    		}
    		if err := s.Err(); err != nil {
    			b.Fatal(err)
    		}
    	}
    	b.Logf("rule: %v", out)
    }
    ```

</details>

But for some reason, `strings.Fields()` is... fastest?

```console
go test -bench=. -benchmem -run XXX -v
goos: darwin
goarch: amd64
pkg: foo
cpu: Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
BenchmarkParseFields-8      	  510610	      1994 ns/op	    4800 B/op	      24 allocs/op
BenchmarkParseFieldFunc-8   	  506881	      2286 ns/op	    4800 B/op	      24 allocs/op
BenchmarkParseSplit-8       	  465390	      2182 ns/op	    4800 B/op	      24 allocs/op
PASS
ok  	foo	3.277s
```


That said, `strings.FieldsFunc()` _WOULD_ allow us to do all splitting at once (including `major:minor`); https://play.golang.org/p/klnqzHEJchG

```go
package main

import (
	"bufio"
	"fmt"
	"strings"
)

func main() {
	list := `c *:* m
b *:* m
c 1:3      rwm
c     1:5 rwm
c 1:7 rwm
c 1:8 rwm
c 1:9 rwm
c 5:0 rwm
c 5:2 rwm
c 136:* rwm
c 10:200 rwm`

	var out []string
	s := bufio.NewScanner(strings.NewReader(list))
	for s.Scan() {
		line := s.Text()
		out = strings.FieldsFunc(line, splitFn)
		fmt.Printf("%+v\n", out)
	}
	if err := s.Err(); err != nil {
		fmt.Println(err)
	}
}

func splitFn(r rune) bool {
	return r == ' ' || r == ':'
}
```

Which produces:

```
[c * * m]
[b * * m]
[c 1 3 rwm]
[c 1 5 rwm]
[c 1 7 rwm]
[c 1 8 rwm]
[c 1 9 rwm]
[c 5 0 rwm]
[c 5 2 rwm]
[c 136 * rwm]
[c 10 200 rwm]
```

Reply:
Giving that idea a try (patch below);

<details>

```patch
diff --git a/libcontainer/cgroups/devices/devices_emulator.go b/libcontainer/cgroups/devices/devices_emulator.go
index aea0f85e..6345e83b 100644
--- a/libcontainer/cgroups/devices/devices_emulator.go
+++ b/libcontainer/cgroups/devices/devices_emulator.go
@@ -78,24 +78,23 @@ func (e *Emulator) IsAllowAll() bool {
 	return e.IsBlacklist() && len(e.rules) == 0
 }
 
+func splitFn(r rune) bool {
+	return r == ' ' || r == ':'
+}
+
 func parseLine(line string) (*deviceRule, error) {
-	// The three fields are: node major:minor perms.
-	fields := strings.Fields(line)
-	if len(fields) != 3 {
+	// The four fields are: node major minor perms.
+	fields := strings.FieldsFunc(line, splitFn)
+	if len(fields) != 4 {
 		return nil, fmt.Errorf("malformed device.list rule %s", line)
 	}
-	// Parse the major:minor pair.
-	mm := strings.Split(fields[1], ":")
-	if len(mm) != 2 {
-		return nil, fmt.Errorf("unexpected minor:major pair %s", fields[1])
-	}
 
 	var (
 		rule  deviceRule
 		node  = fields[0]
-		major = mm[0]
-		minor = mm[1]
-		perms = fields[2]
+		major = fields[1]
+		minor = fields[2]
+		perms = fields[3]
 	)
 
 	// Parse the node type.
```

</details>

Before:

    BenchmarkParseLine-4   	  207354	      5622 ns/op	    5696 B/op	      64 allocs/op

After

    BenchmarkParseLine-4   	  201634	      5288 ns/op	    5520 B/op	      53 allocs/op

Reply:
Indeed, I was mistaken and in this scenario `strings.Split` is faster (not sure why I thought otherwise yesterday, but I discount that to the ongoing heat wave and the lack of AC at home 😓).

I like the idea of using strings.SplitFunc to avoid a second split. Implemented.

Reply:
> I like the idea of using strings.SplitFunc to avoid a second split. Implemented.

Yes! I was drafting the change, and suddenly thought; "hey, we can do it all at once!"